### DISCUSSION TITLE
salmon alignment mode

### DISCUSSION THREAD

**Discussion by aswathysheena** [COMBINE-lab/salmon#887] [Q&A (bulk specific)]
HI,

Hope this is the right forum for this question. 

I have a very simple query, If I use another aligner like STAR, then do run salmon in "alignment" mode, would it perform EM based method to give the trascript estimate or I have start from salmon aligner itself to do EM run?
Thanks,
Aswathy

**Comment by rob-p** [comment]
You can use salmon's EM with the alignments produced by STAR. The only caveat here is that salmon expects alignments to be with respect to the transcriptome rather than the genome.  So you will have to ask STAR to output a transcriptome BAM file instead of or in addition to it's genomic BAM file.  However, that's just a matter of passing STAR the right flags (specifically it's `--quantMode` flag).

**Reply by aswathysheena** [reply]
Thank you very much for the quick response. Follow up to this query, I forgot to mention that I am counting multi-mapped  reads in the STAR using "--outFilterMultimapNmax 100 --winAnchorMultimapNmax 200 ", where outFilterMultimapNmax is the number loci one read can mapped to. (Reason is I have repeated close homologous sequences in my references)

I believe that would smartly handled by the EM method in "alignment" mode and get the quantification? is that correct?

**Reply by rob-p** [reply]
Yes; that's correct.

**Reply by ShwetaCh** [reply]
Hi Rob, 

Thanks for the above insights! I use the same workflow, and so I've more questions in the context of the above conversation. I'm wondering how does Salmon work with the STAR multi-mapping parameters upstream: 

1. Wouldn't the STAR alignment in the bam file be "frozen" in some sense when it is supplied to Salmon as input, and for Salmon to apply its EM algorithm on the transcriptome bam for the purpose of quantitation-only, it would then have to then change the alignments within the bam fundamentally to provide EM-based quantitation, no?

2. Would it do this by ignoring the previous handling of multi-mapping by STAR or taking what it has by STAR and applying its quant level EM on top of it as the next layer? (Given that a bam file is defining the alignment of reads to reference transcriptome. EM is a strategy to handle multi-mapping of reads. Would this not be a conflict between two multi-mapping strategies?) Even more at the conceptual level, I view EM as an algorithm that applies to alignment but not quant. However, it seems that the EM is applicable at both steps - alignment AND quant. Is this correct?

3. What would be the difference between this approach (STAR with multi-mapping parameters + EM) vs. Salmon alone from the start to do EM for alignment and quantitation?

Thanks and apologies in advance for such a meticulous question. Your response is highly appreciated!

**Reply by rob-p** [reply]
Hi @ShwetaCh,

  I think there may be some confusion in terms of what it means for `STAR` to "handle" multi-mapping, versus what `salmon` does.  `STAR` is simply an aligner (an incredibly versatile and damned good one, which is no easy feat) — it's goal is to determine the loci in the genome/transcriptome that best match each read, and at those loci to compute the optimal alignment of the read to the reference.  At that point, `STAR` has completed its job.  Say that for a given read $r$, it has computed all equally-best alignments for $r$ (or, perhaps all optimal alignments and a few slightly sub-optimal alignments).  That's it — `STAR` reports these alignments in the BAM file and no more.

  While this is the ending point for `STAR`, it is only the starting point for `salmon`.  Consider our read $r$, where `STAR` has computed and reported, say, 10 high quality alignments.  These are all *possible* loci that gave rise to the sequenced fragment $r$, but from which locus was $r$ actually derived. After all, even though $r$ may align (even equally well) to 10 different places, this *specific* read actually came from one and only one place.  Well, unfortunately, assigning a distinct, exact, and unambiguous locus to each multi-mapping read is not, in general possible.  So, `salmon` (and similar quantification tools) do the next best thing, they compute a well-defined probability for each possible alignment of the read $r$, and allocate that read proportionally according to these probabilities.  A key insight is that, in isolation, it's difficult to give informative probabilities to the different alignments for $r$, but in the presence of all of our other reads and alignments, there is actually a lot we can say.

  Consider this very simple example.  Imagine $r$ aligns equally well to 10 different transcripts $t_1$, $\dots$, $t_{10}$.  We can't say much in isolation about which alignment is most likely to be the correct one.  But now, consider that in our sample, among all of the other aligned reads, there are 500 reads that align uniquely to $t_{3}$, and no reads, apart from $r$, that align to any of $t_1$ - $t_{10}$.  Now, it seems likely that, among the possible alignments of $r$, the alignment of $r$ to $t_{3}$ is likely to be the correct one.  The other alignments arise and are reported by `STAR` due to sequence similarity, but there is little / no other evidence to support transcripts other than $t_{3}$ being expressed.  Obviously, this is a very simple scenario, but it highlights the intuition that underlies the model and approach.

  So, the way you should think about it is that `STAR` computes the alignments for each read — and for multi-mapping reads, there are multiple *possible* alignments (i.e. multiple loci that could describe the read).  Given these set of possibilities, `salmon` applies statistical inference in a coherent, sample-wide probabilistic model to determine the probabilities that should be associated with each such alignment, and, simultaneously, the most likely abundances of the transcripts that would have given rise to the observed multi-mapping patterns.

Hopefully this helps!

**Comment by ShwetaCh** [comment]
Hi Rob, 

Thank you so much for such a quick response!

The example you described makes it super clear. Now it all makes sense to me - as to how EM applies at the quant level given the set of possiblities of alignments by STAR upstream and they are not at all conflicting and in fact they serve a different role in the process.

Thanks again, I appreciate your time for a detailed response.

Best,
Shweta
